### PR TITLE
Update to spec: Remove overflow: auto/scroll from list of prohibited CSS props

### PR DESCRIPTION
**Pull request by pbakaus**
What the title says.

/cc @cramforce

### CODE REVIEW

**Comment by cramforce** [comment]
I'd rather leave this in place until we actually can document how to do it.

@dvoytenko @camelburrito What is the latest? Do we need a class like `amp-swipable` to get the correct touch semantics working?

**Comment by camelburrito** [comment]
@cramforce  we dont have `amp-swipable`. We usually recommend using `<amp-carousel type="carousel">` for now.

**Comment by pbakaus** [comment]
Ha, I wasn't aware that overscroll scrolling is completely blocked. Pretty unfortunate for our code blocks: https://www.google.com/amp/s/www.ampproject.org/docs/guides/responsive/art_direction.html

**Comment by pbakaus** [comment]
So I think we need to make it work, automagically. It's possible to detect overflowing containers, and do some nifty JS to translate the scroll.

**Comment by dvoytenko** [comment]
I believe we just need to add `amp-scrollable` element. There are too many nuances and performance is too critical for scrollable elements, thus we can't scan DOM for `overflow` CSS.

**Comment by camelburrito** [comment]
Things that we need to think about
1) How to load/not load AMP elements that are visible/hidden in/from the viewport
2) How to keep the document swipe-able.
3) Should we support `overflow-x` only?

**Comment by pbakaus** [comment]
I don't buy that argument yet :)

You don't need to scan the DOM and parse CSS. All you need is to test for one of the two:

1. Get all DOM elements, scan from innermost to outermost and find parentElem.scrollParent
2. or: scan through all elements, and compare scrollWidth with clientWidth

And frankly, I'd do this on demand anyway:

1. Touch registers (`touchstart` or `mousedown`)
2. Quickly look at event.target.scrollWidth (and check parents too)
3. Do the appropriate thing.

I've actually done this before.

**Comment by cramforce** [comment]
We're not doing such scans in principle. You also have to redo them when
CSS changes at runtime.

**Comment by pbakaus** [comment]
OK let's leave the big scans – the "on demand" solution is IMHO still valid, and very cheap in practice. I simply encourage you to try hard to solve this problem without inventing another component you need to support for eternity.

We get enough criticism for "reinventing the wheel", and I'd argue that yes, for probably the first time, we actually would here, for less good reasons than elsewhere. We're after all talking about this because of an implementation detail of how the *AMP Viewer* works, not the runtime :)

**Comment by cramforce** [comment]
Yeah, I'm not in favor of a component. The first thing would be a class
name that would unbreak users of overflow

**Comment by dvoytenko** [comment]
@pbakaus @cramforce for now I see it this way:

> and find parentElem.scrollParent
> or: scan through all elements, and compare scrollWidth with clientWidth

What is `scrollParent`? Finding scroll parents in reality is hugely expensive. Doing so on every `touchstart` is also very expensive.

> The first thing would be a class name that would unbreak users of overflow

At this point, I don't really see difference between a class and new element name. In particular, our CSS classes are harder to document and control. I'd much rather have a stricter validator rules and an element to make it clear what's expected.

The bottom line, there are a lot of things that need to be done under the hood for horizontal scrolling to work with any reasonable UX. And embed-ability is a big feature of AMP.

**Comment by pbakaus** [comment]
> What is scrollParent? Finding scroll parents in reality is hugely expensive. Doing so on every touchstart is also very expensive.

My bad, scrollParent is a function I wrote for jQuery UI at some point: https://github.com/slindberg/jquery-scrollparent/blob/master/jquery.scrollparent.js

Still don't buy the expensive part. `scrollWidth === clientWidth` should be fast enough, and should produce just one layout (when checking it at the same time for the parent chain). Maybe I'll produce a quick test to prove it :)

> The bottom line, there are a lot of things that need to be done under the hood for horizontal scrolling to work with any reasonable UX. And embed-ability is a big feature of AMP.

Horizontal scrolling has been a feature of the web for a long time. All browsers handle it somewhat well with appopriate UX. It works. It would be quite sad if we can't make it work in AMP.

**Comment by dvoytenko** [comment]
@pbakaus 

> Still don't buy the expensive part. scrollWidth === clientWidth should be fast enough

Sure. Let's look at the test. But please consider that this test will essentially have to be performed for all touchstart/touchmove events for an arbitrary number of DOM elements. I can't imagine this ever working fast enough. Compare it with a simple implementation of a hypothetical AMP element where we'd simply need to stop propagation on a couple of events.

> Horizontal scrolling has been a feature of the web for a long time. All browsers handle it somewhat well with appopriate UX. It works. It would be quite sad if we can't make it work in AMP.

Sure. But we need to find a way where it actually works. Let's keep in mind that there's a big set of things that are needed here besides enabling document-swipe in viewers, such as:

- All CSS should be set correctly with things such as `-webkit-overflow-scrolling: touch` for iOS.
- It's possible that we'd need `touch-action: pan-x` or similar customization.
- We need protection from iOS scroll freezing. See #330 as an example. Our carousels provide a lot of workarounds for this.
- We need to connect visibility to resource loader to pre-render/render elements scrolled into view.

**Comment by pbakaus** [comment]
Argh I don't have time to build a test in the next few days. Please feel free to take that action item from me, or I'll resume when I have some room to breathe after CDS.

**Comment by pbakaus** [comment]
Bot reminds me this is a good time to reiterate that I think this is a problem worth solving. Not supporting overflow might hinder adoption of canonical AMP.

**Comment by aghassemi** [comment]
Thanks for the pull request. Since this PR has been stale for a while, I am closing it. Please feel free to reopen as needed.
